ddl/multi_schema_change.go

```diff
@@ -323,6 +326,41 @@ func checkOperateSameColAndIdx(info *model.MultiSchemaInfo) error {
 	return checkIndexes(info.AlterIndexes, true)
 }
 
+func checkOperateDropIndexUseByForeignKey(info *model.MultiSchemaInfo, t table.Table) error {
+	var remainIndexes, droppingIndexes []*model.IndexInfo
+	tbInfo := t.Meta()
+	for _, idx := range tbInfo.Indices {
+		dropping := false
+		for _, name := range info.DropIndexes {
+			if name.L == idx.Name.L {
+				dropping = true
+				break
+			}
+		}
+		if dropping {
+			droppingIndexes = append(droppingIndexes, idx)
+		} else {
+			remainIndexes = append(remainIndexes, idx)
+		}
+	}
+
+	for _, fk := range info.AddForeignKeys {
+		if droppingIdx := findIndexByColumns(tbInfo, droppingIndexes, fk.Cols...); droppingIdx != nil && findIndexByColumns(tbInfo, remainIndexes, fk.Cols...) == nil {
+			return dbterror.ErrDropIndexNeededInForeignKey.GenWithStackByArgs(droppingIdx.Name)
+		}
+	}
+	return nil
+}
+
+func findIndexByColumns(tbInfo *model.TableInfo, indexes []*model.IndexInfo, cols ...model.CIStr) *model.IndexInfo {
+	for _, index := range indexes {
+		if model.IsIndexPrefixCovered(tbInfo, index, cols...) {
+			return index
+		}
+	}
+	return nil
+}
```

Comment:
Could you move this to `parser/model/model.go` and update `model.FindIndexByColumns`?

Reply:
done.